security/pkg/nodeagent/sds/sdsservice_test.go

```diff
@@ -0,0 +1,177 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package sds
+
+import (
+	"fmt"
+	"net"
+	"reflect"
+	"testing"
+	"time"
+
+	api "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	authapi "github.com/envoyproxy/go-control-plane/envoy/api/v2/auth"
+	core "github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
+	sds "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v2"
+	"github.com/gogo/protobuf/types"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+	"k8s.io/apimachinery/pkg/util/uuid"
+)
+
+var (
+	socket               string
+	fakeCertificateChain = []byte{01}
+	fakePrivateKey       = []byte{02}
+)
+
+func TestStreamSecrets(t *testing.T) {
+	socket = fmt.Sprintf("/tmp/gotest%q.sock", string(uuid.NewUUID()))
+	testHelper(t, socket, sdsRequestStream)
+}
+
+func TestFetchSecrets(t *testing.T) {
+	socket = fmt.Sprintf("/tmp/gotest%q.sock", string(uuid.NewUUID()))
+	testHelper(t, socket, sdsRequestFetch)
+}
+
+type secretCallback func(string, *api.DiscoveryRequest) (*api.DiscoveryResponse, error)
+
+func testHelper(t *testing.T, testSocket string, cb secretCallback) {
+	arg := Args{
+		SDSUdsSocket: testSocket,
+	}
+	st := &mockSecretStore{}
+	server, err := NewServer(arg, st)
+	defer server.Close()
+
+	if err != nil {
+		t.Fatalf("failed to start grpc server for sds: %v", err)
+	}
+
+	req := &api.DiscoveryRequest{
+		ResourceNames: []string{"test"},
+		Node: &core.Node{
+			Id: "sidecar~127.0.0.1~id~local",
+		},
+	}
+
+	wait := 300 * time.Millisecond
+	for try := 0; try < 5; try++ {
+		time.Sleep(wait)
+		// Try to call the server
+		resp, err := cb(testSocket, req)
+		if err == nil {
+			//Verify secret.
+			var pb authapi.Secret
+			if err := types.UnmarshalAny(&resp.Resources[0], &pb); err != nil {
+				t.Fatalf("UnmarshalAny SDS response failed: %v", err)
+			}
+
+			certificateChainGot := pb.GetTlsCertificate().GetCertificateChain()
+			certificateChainWant := &core.DataSource{
+				Specifier: &core.DataSource_InlineBytes{
+					InlineBytes: fakeCertificateChain,
+				},
+			}
+			if !reflect.DeepEqual(certificateChainWant, certificateChainGot) {
+				t.Errorf("certificate Chain: got %+v, want %+v", certificateChainGot, certificateChainWant)
+			}
+
+			privateKeyGot := pb.GetTlsCertificate().GetPrivateKey()
+			privateKeyWant := &core.DataSource{
+				Specifier: &core.DataSource_InlineBytes{
+					InlineBytes: fakePrivateKey,
+				},
+			}
+			if !reflect.DeepEqual(privateKeyWant, privateKeyGot) {
+				t.Errorf("private key: got %+v, want %+v", privateKeyGot, privateKeyWant)
+			}
+
+			return
+		}
+
+		wait *= 2
+	}
+
+	t.Fatalf("failed to start grpc server for SDS")
```

Comment:
This error message is misleading. Could it be the for loop retries all failed 5 times in a row, and then reach it here?

And the current implementation already starts the gRPC server in `NewServer`.

Reply:
It's possible, the reason of adding for loop here is server start in a go routine(NewServer->initDiscoveryService->grpcServer.Serve), which means the server may not be ready if we send request immediately after calling NewServer, so retry here give some time to make sure server is ready to serve traffic (and failed the test if not ready for some time).